crates/matrix-sdk/src/latest_events/latest_event.rs

```diff
@@ -150,6 +169,332 @@ impl LatestEventValue {
             .map(Self::Remote)
             .unwrap_or_default()
     }
+
+    /// Create a new [`LatestEventValue::LocalIsSending`] or
+    /// [`LatestEventValue::LocalIsWedged`].
+    async fn new_local(
+        send_queue_update: &RoomSendQueueUpdate,
+        buffer_of_values_for_local_events: &mut LatestEventValuesForLocalEvents,
+        room_event_cache: &RoomEventCache,
+        power_levels: &Option<(&UserId, RoomPowerLevels)>,
+    ) -> Self {
+        use crate::send_queue::{LocalEcho, LocalEchoContent};
+
+        match send_queue_update {
+            // A new local event is being sent.
+            //
+            // Let's create the `LatestEventValue` and push it in the buffer of values.
+            RoomSendQueueUpdate::NewLocalEvent(LocalEcho {
+                transaction_id,
+                content: local_echo_content,
+            }) => match local_echo_content {
+                LocalEchoContent::Event { serialized_event: content, .. } => {
+                    if let Ok(content) = content.deserialize() {
+                        if let Some(kind) = find_and_map_any_message_like_event_content(content) {
+                            let value = Self::LocalIsSending(kind);
+
+                            buffer_of_values_for_local_events
+                                .push(transaction_id.to_owned(), value.clone());
+
+                            value
+                        } else {
+                            Self::None
+                        }
+                    } else {
+                        Self::None
+                    }
+                }
+
+                LocalEchoContent::React { .. } => Self::None,
+            },
+
+            // A local event has been cancelled before being sent.
+            //
+            // Remove the calculated `LatestEventValue` from the buffer of values, and return the
+            // last `LatestEventValue` or calculate a new one.
+            RoomSendQueueUpdate::CancelledLocalEvent { transaction_id } => {
+                if let Some(position) = buffer_of_values_for_local_events.position(transaction_id) {
+                    buffer_of_values_for_local_events.remove(position);
+                }
+
+                Self::new_local_or_remote(
+                    buffer_of_values_for_local_events,
+                    room_event_cache,
+                    power_levels,
+                )
+                .await
+            }
+
+            // A local event has successfully been sent!
+            //
+            // Unwedge all wedged values after the one matching `transaction_id`. Indeed, if
+            // an event has been sent, it means the send queue is working, so if any value has been
+            // marked as wedged, it must be marked as unwedged. Then, remove the calculated
```

Comment:
:bug: The second sentence starting with `Indeed` shows "spooky action at a distance" tight coupling. On the one hand, it's bad in terms of design, on the other hand, it's likely wrong in the sense that "wedged" means there was a fatal error and there's no way a retry without updating the event could fix it. So values marked as wedged shouldn't be marked unwedged, in this case.